### PR TITLE
UI refactor, take 2

### PR DESCRIPTION
**Pull request by kingoflolz**

### CODE REVIEW

**Comment by aeickhoff** [comment]
Can I help with getting this working? Would be cool to have it in the next release!

**Comment by kingoflolz** [comment]
It is almost working at the moment, all controls work, but the things that still need to be done are:
- Saving and loading of settings (also naming of settings)
- Fix a small bug with hovering
- Finish exclusive events

**Comment by kingoflolz** [comment]
Small update, exclusive events are working now(tested with redo/undo) but I have introduced some weird bugs in the process...

**Comment by kingoflolz** [comment]
I resolved the hover issue, but to be honest, I have no idea how it ever worked before haha

**Comment by kingoflolz** [comment]
Note: you need to click somewhere on the map to start a grid, to make the canvas the active interactable.
Spent 30 mins debugging this T.T

**Comment by kingoflolz** [comment]
All that's left for tomorrow is saving and cleanup!

**Comment by aeickhoff** [comment]
Can't you manually send user_interface::Focus in the beginning, or does that not exist anymore?

**Comment by kingoflolz** [comment]
Yeah, tomorrow I'll initialise the user interface struct with the active interactable being the canvas

**Comment by aeickhoff** [comment]
Hey @kingoflolz: I'm almost done with my planning refactor and it would make a lot of sense if I just merged this pull request into that one and did the hooking-up myself in the new code (it's pretty different from the old one). Then if anything doesn't work in your new code, you can help me resolve it in my PR. How does that sound?

**Comment by kingoflolz** [comment]
@aeickhoff Sure!

**Comment by aeickhoff** [comment]
I'll solve the conflicts and do the hooking up, I'll let you know how it goes

**Comment by aeickhoff** [comment]
Two problems:
- [ ] KeyUp/MouseUp seems to be not implemented at all (required for dragging)
- [ ] The code is quite messy

I will address both of them here before merging and try to document my changes as feedback for @kingoflolz

**Comment by aeickhoff** [comment]
Unfortunately after digging slightly deeper, trying to implement handling of releasing keys, I have to come to the following conclusion:

**This code, although it contains good ideas, is too messy from the ground up!**

Big problems:

- Responsibilities for handling events are all over the place
  - partially in `ui`
  - partially in `user_interface`
  - partially in `settings`
- Any incoming event goes through all of these and is untransparently transformed into `KeyOrButton` and then later `Action::KeyHeld/Up/Down` with just an `action_id` or `Action::Mouse` which contains both and `action_id` a `Mouse::Moved/Scrolled/Up/Down`... what??
- Generic user interface event dispatching and camera controls are still mixed, with the user interface basically being its own listener in a schizophrenic way
- Code style that makes it very hard to follow what's happening
  - using tuples instead of more helpful custom types with named fields
  - not even destructuring tuples into named variables, but doing stuff like `tuple.0`
- UI state that is somehow held only in the main function?! (`keys_down`)

Smaller issue which I wanted to address in the future anyways

- The settings serialisation with serde seems very over-engineered and still not human-friendly

**New Plan**

I will reimplement all of this in a cleaner way, keeping your good ideas (e.g. key combo registration). Please don't take this as an insult @kingoflolz, I just have to draw a line where something causes me too much work to fix vs. redo it. I really appreciate your work and I hope you will be able to find your ideas again in my implementation!

**Comment by kingoflolz** [comment]
@aeickhoff I completely respect your decision on the issue, as it is your repository and for you to maintain. I think that I have made some *interesting* (probably wrong in hindsight) decisions which made the code more complex than it needs to be, and almost everything was written in a rush, and with a headache, so it is definitely not the best code (the lack of comments did not really help either). I was planning to clean it up after I have completed saving, so it wasn't that great of a time for you to look at it haha.
What I intended was for user_interface to be the camera and clicking/dragging dispatching, settings to just store settings and ui to be renamed to input_dispatcher and have some methods from settings pulled into functions there.
Interested to read how you manage to do it!

**Comment by aeickhoff** [comment]
Thank you for your understanding and for your clarifications! I already got the impression that I had misguided expectations about the state of the implementation, but I still think that in general a rewrite is better than trying to base it on existing structures.
